Android-NDK/jniapi.cpp

```diff
@@ -55,6 +69,28 @@ jobject wrapNode(JNIEnv* jenv, NodePointer node) {
     return wrapper;
 }
 
+//helper function for wrappers returning a traceroute probe result
+jobject wrapProbe(JNIEnv* jenv,
+                  std::string probeAddress,
+                  bool success,
+                  double elapsedMs) {
+
+    //strings that need to be freed after
+    jstring from = jenv->NewStringUTF(probeAddress.c_str());
+    jclass probeWrapperClass = jenv->FindClass("com/peer1/internetmap/ProbeWrapper");
+
+    jmethodID constructor = jenv->GetMethodID(probeWrapperClass, "<init>", "(ZLjava/lang/String;D)V");
+    //note: if you change this code, triple-check that the argument order matches ProbeWrapper.
+    probeWrapper = jenv->NewObject(probeWrapperClass, constructor, success, from, elapsedMs);
```

Comment:
Can I accomplish this by simply setting probeWrapper to a nullptr before setting it again?

Reply:
No. C++ has no automatic memory management, so seeing a pointer to NULL won't ever do anything, if it needs to be manually freed, your re always going to have to call some delete function. 

I THOUGHT you'd need need to manually call DeleteGlobalRef on it (that's what it is doing later on to free it when everything is cleaned up). However, looking at the documentation for how local and global refs work (https://docs.oracle.com/javase/6/docs/technotes/guides/jni/spec/design.html#wp1242), I'm wondering if actually the bug here ISN'T that the memory is leaked, but that the value stored in probeWrapper is NOT actually legitimately referenced, and so could be freed and go invalid, and that's just never happening in practice. It doesn't look like it ever even uses it. So most likely there isn't actually any fix needed for correctness, and the right fix to make it clearer what is going on is just to not have that global reference to probeWrapper, and delete the existing (i think incorrect) DeleteGlobalRef on it.

Reply:
@nbrooke Originally I had created a local reference for ProbeWrapper, allocated and deleted it with each call. This seemed to be causing large memory allocation spikes when I was running performance tests on the feature. In some cases, I was seeing crashes due to memory allocation issues (although this may be an OS specific issue, I wrote up some details in the original issue for this).

I was attempting to reduce these spikes by re-using a single, Global instance of the wrapper.

Do you have any ideas on how I can improve the memory spikes if I switch back to local refs?

Reply:
Unless i'm missing something, the current code is not re-using a global instance. It stores the previous probe wrapper instance, but never seems to reuse it (next time probeDestinationAddressWithTTL is called, it just calls wrapProbe again which allocates a new instance. That's why I think nothing bad is happening here, it store probe wrapper in a global variable, but doesn't call NewGlobalRef on it (which I believe it should be if it really wanted to do that) so that object will actually get freed as soon as the java code that is receiving it from probeDestinationAddressWithTTL stops referencing it.